AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -41,70 +42,81 @@ @interface ASVideoNode ()
 
 @implementation ASVideoNode
 
+//TODO: Have a bash at supplying a preview image node for use with HLS videos as we can't have a priview with those
+
+
+#pragma mark - Construction and Layout
+
 - (instancetype)init
 {
-  if (!(self = [super init])) {
-    return nil;
-  }
-    
-  _previewQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0);
+	_previewQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0);
   
   self.playButton = [[ASDefaultPlayButton alloc] init];
-  
   self.gravity = AVLayerVideoGravityResizeAspect;
-  
   [self addTarget:self action:@selector(tapped) forControlEvents:ASControlNodeEventTouchUpInside];
-    
+  
   return self;
 }
 
-- (void)interfaceStateDidChange:(ASInterfaceState)newState fromState:(ASInterfaceState)oldState
+- (instancetype)initWithViewBlock:(ASDisplayNodeViewBlock)viewBlock didLoadBlock:(ASDisplayNodeDidLoadBlock)didLoadBlock
 {
-  [super interfaceStateDidChange:newState fromState:oldState];
-  
-  if (!(newState & ASInterfaceStateVisible)) {
-    if (oldState & ASInterfaceStateVisible) {
-      if (_shouldBePlaying) {
-        [self pause];
-        _shouldBePlaying = YES;
-      }
-      [(UIActivityIndicatorView *)_spinner.view stopAnimating];
-      [_spinner removeFromSupernode];
-    }
-  } else {
-    if (_shouldBePlaying) {
-      [self play];
-    }
-  }
+  ASDisplayNodeAssertNotSupported();
+  return nil;
 }
 
-- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+- (ASDisplayNode*)constructPlayerNode
 {
-  if ([change[@"new"] integerValue] == AVPlayerItemStatusReadyToPlay) {
-    if ([self.subnodes containsObject:_spinner]) {
-      [_spinner removeFromSupernode];
-      _spinner = nil;
+  ASDisplayNode* playerNode = [[ASDisplayNode alloc] initWithLayerBlock:^CALayer *{
+    AVPlayerLayer *playerLayer = [[AVPlayerLayer alloc] init];
+    if (!_player) {
+      [self constructCurrentPlayerItemFromInitData];
+      _player = [AVPlayer playerWithPlayerItem:_currentPlayerItem];
+      _player.muted = _muted;
     }
-  }
+    playerLayer.player = _player;
+    playerLayer.videoGravity = [self gravity];
+    return playerLayer;
+  }];
   
-  if ([change[@"new"] integerValue] == AVPlayerItemStatusFailed) {
-    
+  return playerNode;
+}
+
+- (void)constructCurrentPlayerItemFromInitData
```

Comment:
This can be called multiple times.  We should probably skip the work if _currentPlayerItem exists; should not replace it.  Maybe check if its asset is _asset.